src/optimise/train.jl

```diff
@@ -97,12 +102,13 @@ Multiple optimisers and callbacks can be passed to `opt` and `cb` as arrays.
 function train!(loss, ps, data, opt; cb = () -> ())
   ps = Params(ps)
   cb = runall(cb)
+  st = [Optimisers.init(opt, p) for p in ps]
```

Comment:
Could we define `init` for `Params` like we do `update`?

Reply:
I'm thinking if this should be wholesale replaced with `Optimisers.state` instead

Reply:
I'm only worried about corner cases that this might hit

Reply:
I think `state` makes sense. For every `p` in `ps`, we should be able to independently optimize each one by calling the user facing `state` and `update`. I think its safe to do that here.

Reply:
Right, so that works in cases where we have arrays as params, but not arbitrary structs, which is what we want.

The user facing API is something like (pseudocode; although it might work in this MWE)
```julia
m = Dense(3,3)
opt = ADAM()
st = Optimisers.state(opt, m) # `m` could contain arbitrary structs which shouldn't be functor'd
loss(m, x, y) = Flux.mse(m(x), y)
for i = 1:1000
  gs, = gradient(m) do m
    @show loss(m, w, w′)
  end
  m, st = opt(m, gs, st)
end
```

Reply:
That's exactly what I'm suggesting. To be more specific, change the code to:
```suggestion
  st = state(opt, ps)
```

And elsewhere, define:
```julia
init(o, ps::Params) = [init(o, p) for p in ps]
```

Should allow for the future case where `m` is not a `Params` but also the current case where we need to support `Params`.